_posts/2021-07-13-JEV.md

```diff
@@ -0,0 +1,822 @@
+instruction. While the speed gain is usually negligible, your code becomes smaller by saving the additional bytes[2].
+
+#### -- Interlude --
+
+So again, forgive the long-winded introduction, but as usual, it's easier to relate points to each other.
+
+If we reduce the fact that SIMD instructions are simply (fancy)CPU instructions, and add that to the fact that an
+Optimizing Compiler chooses the best `opcode/instruction` for the job, then we might conclude that the Compiler might
+also choose a SIMD instruction, if the situation arises. And we would be correct.
+
+And this process my dear comrades, is called: Auto Vectorization.
+
+![Dramatic look](https://media.giphy.com/media/kKdgdeuO2M08M/giphy.gif)
+
+Does this mean my trustworthy compiler will Vectorize all my code?
+As a wise man once said...it depends.
+
+To explain this, we will need to quote a different wise man, who once said that compilers are basically "stupid regex
+pattern matching engines" (sorry, can't remember the source). Meaning, like everything your compiler does, it only does
+so under very specific circumstances. So while the magical compiler might rewrite the above loop to use SIMD, it doesn't
+take a lot of deviation for the same wizardly compiler to skip it.
+
+Non-SIMD example:
+
+ ```java
+  for (int i = 0; i < 6220800; i++) {
+      if (i > 0)
+          rgb[i]++;
+  }
+```
+
+Granted that you wouldn't write code like this (on purpose), it is still baffling that something like this would keep us
+from our deserved Vectorization goodness ¯\\_(ツ)_/¯
+
+#### What abouts the Java?
+
+Okay then, back to Java reality with the million dollar question; ~~to be or not to be~~ what about Java? Well,
+considering that Java compilers don't actually compile to Assembly/Machine Code T_T, we are sorry to say that...the JVM
+architects/developer got you covered and already implemented this in the JVM portion for your pleasure (we're not worthy).
+
+So yeah, Java supports a substantial subset of Auto Vectorization[3]. But like everything in JavaLand that concerns
+real-life hardware, there is an added layer of indirection that needs to be taken into consideration.
+
+If you remember, the Java code lifecycle towards the CPU is an arduous one:
+
+- Interpreter
+- C1
+- C2
+
+As such, the AV optimizations are usually only activated from the latter part of C1 onwards (downwards?).
+
+Meaning (among other things), that very short lived pieces of code that don't reach C2 would unfortunately not be
+considered AV candidates. This on top of the compiler level requirements for a piece of code (such as a simple loop) to
+be Vectorizable.
+
+### Explicit Vectorization (EV)
+
+#### Prelude
+
+We've always been taught the magic words `"the compiler is smart enough to..."`, which in turn would mean that if a
+compiler fails to do whatever it is we expect of it, then the 'fault' lies with layer 8.
```

Comment:
Layer 8, very nais.

So from now on: 'Problem exists between keyboard and layer 8'